File: ocaml/hash-map.ml
(* Std-collection workload on OCaml's standard hash map
   (Stdlib.Hashtbl, mutable in place, seeded polynomial hash). Large
   and broad: keys are raw MINSTD draws, uniform over [1, 2^31-2];
   MINSTD is a full-period permutation, so fresh draws never repeat,
   and removals plus the hit half of the lookups replay the build key
   stream through a second MINSTD state. Build 4M inserts, churn 2M
   rounds, 2M lookups alternating replayed hits and fresh misses.
   Final size 4,999,834; the checksum is iteration-order independent so
   all representations agree. Replace semantics keep at most one
   binding per key. *)

let build_n = 4_000_000
let churn_n = 2_000_000
let lookup_n = 2_000_000
let expected = 166_401_892_080_070_584L
let lcg x = Int64.rem (Int64.mul x 48_271L) 2_147_483_647L

let () =
  let m : (int64, int64) Hashtbl.t = Hashtbl.create 1024 in
  let x = ref 1L in
  for i = 0 to build_n - 1 do
    x := lcg !x;
    Hashtbl.replace m !x (Int64.of_int i)
  done;
  let r = ref 1L in
  for i = 0 to churn_n - 1 do
    x := lcg !x;
    if Int64.rem !x 4L = 3L then (
      r := lcg !r;
      Hashtbl.remove m !r)
    else Hashtbl.replace m !x (Int64.of_int (build_n + i))
  done;
  let acc = ref 0L in
  for _ = 0 to lookup_n - 1 do
    x := lcg !x;
    let k =
      if Int64.rem !x 2L = 0L then (
        r := lcg !r;
        !r)
      else !x
    in
    let v = match Hashtbl.find_opt m k with Some v -> v | None -> -1L in
    acc := Int64.add !acc v
  done;
  let folded =
    Hashtbl.fold
      (fun k v a -> Int64.add a (Int64.add (Int64.mul k 31L) v))
      m 0L
  in
  let result =
    Int64.add
      (Int64.add !acc folded)
      (Int64.mul 7L (Int64.of_int (Hashtbl.length m)))
  in
  if result <> expected then (
    Printf.eprintf "FAIL: expected %Ld, got %Ld\n" expected result;
    exit 1)

File: haskell/hash-map.hs
{-# LANGUAGE BangPatterns #-}

-- Std-collection workload on Haskell's de-facto standard hash map
-- (Data.HashMap.Strict from unordered-containers — a persistent HAMT,
-- the same representation family as Reussir's std HashMap). Same
-- workload and checksum as ordered-map.hs; the checksum is
-- iteration-order independent so the representations agree.

import Data.Int (Int64)
import qualified Data.HashMap.Strict as HM
import System.Exit (exitFailure)

keyspace, buildN, churnN, lookupN, expected :: Int64
keyspace = 524287
buildN = 1000000
churnN = 1000000
lookupN = 1000000
expected = 105140861851414131

lcg :: Int64 -> Int64
lcg x = (x * 48271) `rem` 2147483647

buildLoop :: Int64 -> Int64 -> HM.HashMap Int64 Int64 -> (Int64, HM.HashMap Int64 Int64)
buildLoop !i !x !m
  | i == buildN = (x, m)
  | otherwise =
      let x' = lcg x
       in buildLoop (i + 1) x' (HM.insert (x' `rem` keyspace) i m)

churnLoop :: Int64 -> Int64 -> HM.HashMap Int64 Int64 -> (Int64, HM.HashMap Int64 Int64)
churnLoop !i !x !m
  | i == churnN = (x, m)
  | otherwise =
      let x' = lcg x
          k = x' `rem` keyspace
          m' =
            if x' `rem` 4 == 3
              then HM.delete k m
              else HM.insert k (buildN + i) m
       in churnLoop (i + 1) x' m'

lookupLoop :: Int64 -> Int64 -> HM.HashMap Int64 Int64 -> Int64 -> Int64
lookupLoop !i !x !m !acc
  | i == lookupN = acc
  | otherwise =
      let x' = lcg x
       in lookupLoop (i + 1) x' m (acc + HM.findWithDefault (-1) (x' `rem` keyspace) m)

main :: IO ()
main = do
  let (x1, built) = buildLoop 0 1 HM.empty
      (x2, m) = churnLoop 0 x1 built
      acc = lookupLoop 0 x2 m 0
      folded = HM.foldlWithKey' (\a k v -> a + k * 1000003 + v) 0 m
      result = acc + folded + 7 * fromIntegral (HM.size m)
  if result == expected
    then pure ()
    else do
      putStrLn ("FAIL: expected " ++ show expected ++ ", got " ++ show result)
      exitFailure
